pkg/apis/walruscore/v1/resource_run.go

```diff
@@ -41,3 +109,191 @@ type ResourceRunList struct {
+
+// ResourceRunStepType describes the type of the resource run step.
+// +enum
+type ResourceRunStepType string
+
+const (
+	// ResourceRunStepTypePlan is the type of the plan resource, it is system generated.
+	ResourceRunStepTypePlan ResourceRunStepType = "Plan"
+	// ResourceRunStepTypeApply is the type of the apply resource, it is system generated.
+	ResourceRunStepTypeApply ResourceRunStepType = "Apply"
+
+	// ResourceRunStepTypeContainer is the type of the container resource, it depends on the resource run hooks and resource run template.
+	ResourceRunStepTypeContainer ResourceRunStepType = "Container"
+	// ResourceRunStepTypeApproval is the type of the approval resource, it depends on the resource run hooks and resource run template.
+	ResourceRunStepTypeApproval ResourceRunStepType = "Approval"
+	ResourceRunStepTypeUnknown  ResourceRunStepType = "Unknown"
+)
+
+// ResourceRunStep stores the step run result of the resource run.
+// The step may be the plan, apply, and hook step defined in the resource hook or resource run template.
+type ResourceRunStep struct {
+	// StatusDescriptor defines the status of the step.
+	StatusDescriptor `json:",inline"`
+
+	// Name is the name of the step.
+	Name string `json:"name"`
+
+	// Type is the type of the step.
+	Type ResourceRunStepType `json:"type"`
+
+	// ResourceRunStepTemplate is a reference to a resource step template.
+	ResourceRunStepTemplate *ResourceRunStepTemplateRef `json:"runStepTemplate,omitempty"`
+
+	// StartedAt is the time when the step started.
+	StartedAt meta.Time `json:"startedAt"`
+
+	// FinishedAt is the time when the step finished.
+	FinishedAt meta.Time `json:"finishedAt"`
+
+	// Duration is the duration of the step.
+	Duration *int64 `json:"duration"`
+
+	// Container is the container configs to run in the resource run step.
+	// Only used when ResourceStepTemplate is not provided.
+	Container *core.Container `json:"container,omitempty"`
+
+	// ApprovalRecord stores the approval records of the step.
+	// Only the step template reference that has approval spec will have approval record.
+	ApprovalRecord *ResourceRunStepApprovalRecord `json:"approvalRecord,omitempty"`
+}
+
+// ResourceRunConfig is the generated configurations of the resource run.
+// It reference to an secret object.
+type ResourceRunConfig struct {
+	SecretRef *core.LocalObjectReference `json:"secretRef"`
+}
+
+// ResourceRunStepApprovalRecord stores the approval records of target resource run step.
+type ResourceRunStepApprovalRecord struct {
+	// ApprovalType is the type of the approval process.
+	Type ResourceRunStepTemplateApprovalType `json:"approvalType"`
+
+	// ApprovalUsers is the users to approve the step.
+	ApprovalUsers []string `json:"approvalUsers"`
+
+	// ApprovedUsers is the users who approved the step.
+	ApprovedUsers []string `json:"approvedUsers,omitempty"`
+
+	// RejectedUsers is the users who rejected the step.
+	RejectedUsers []string `json:"rejectedUsers,omitempty"`
```

Comment:
only one question: why not use one UserActions slice to replace two opposite users slices?

```
type ResourceRunStepApprovalAction string
const (
    ResourceRunStepApproval ResourceRunStepApprovalAction = "Approval"
    ResourceRunStepReject ResourceRunStepApprovalAction = "Reject"
)

type ResourceRunStepApprovalActionRecord {
   User string // the name of the user
   Action ResourceRunStepApprovalAction // action
}

type ResourceRunStepApprovalRecord struct {
  Users []string `json:"users"` // the users to approve the step.
  UserActions []ResourceRunStepApprovalActionRecord `json:"userActions"`
}

func (in ResourceRunStepApprovalRecord) IsApproved() bool {
  if len(in.UserActions) == 0 {
    return false
  }
  for i := range in.UserActions {
    if in.UserActions[i] != ResourceRunStepApproval {
      return false
    }
  }
  return true
}

func (in ResourceRunStepApprovalRecord) IsRejected() bool {
  return !in.Approved()
}

```

Reply:
Updated. The UX need update simultaneously.